runtime/POSIX/fd.c

```diff
@@ -128,47 +129,38 @@ static int has_permission(int flags, struct stat64 *s) {
   gid_t egid = getegid();
 #endif  
 
-  if (read_access && ((mode & S_IRUSR) | (mode & S_IRGRP) | (mode & S_IROTH)))
+  if (read_access && !((mode & S_IRUSR) | (mode & S_IRGRP) | (mode & S_IROTH))) {
+    // klee_warning("No permission for read");
+    fprintf(stderr, "read access: %d, %o, %o\n", read_access, mode, S_IRUSR| S_IRGRP | S_IROTH);
     return 0;
+  }
 
-  if (write_access && !((mode & S_IWUSR) | (mode & S_IWGRP) | (mode & S_IWOTH)))
+  if (write_access && !((mode & S_IWUSR) | (mode & S_IWGRP) | (mode & S_IWOTH))) {
+    // klee_warning("No permission for write");
+    fprintf(stderr, "read access: %d, %o, %o\n", read_access, mode, S_IWUSR| S_IWGRP | S_IWOTH);
     return 0;
+  }
 
   return 1;
 }
 
+static int __concretize_int(int i) {
+  int c = klee_get_value_i32(i);
+  klee_assume(c == i);
+  return c;
+}
 
-// largely copied from fd_init.c:__create_new_dfile
-// biggest difference is not allocating the contents until the mmap later
-// return value: 0 means fine, non-zero means error
-static int __create_pmem_dfile(exe_disk_file_t *dfile, unsigned size,
-                               const char *name, struct stat64 *defaults) {
-  struct stat64 *s = malloc(sizeof(*s));
-  const char *sp;
-  char sname[64];
-  for (sp=name; *sp; ++sp)
-    sname[sp-name] = *sp;
-  memcpy(&sname[sp-name], "-stat", 6);
-
-  assert(size);
-  if (size % 4096 != 0) {
-    klee_error("pmem file size must be multiple of page size (4096)");
-    return -1;
-  }
+static void __setup_pmem_struct_stat(struct stat64 *s, exe_disk_file_t *dfile, struct stat64 *defaults, const char *sname) {
 
-  dfile->size = size;
-  // FIXME: page aligned vs. cache aligned
-  dfile->contents = memalign(4096, dfile->size);
```

Comment:
should this also be pgsz, rather than 4096?

Reply:
Yup.

Reply:
This was already deleted...